File: leetcode/131.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ret;
        vector<string> elm;
        dfs(ret, elm, s, 0);
        return ret;
    }
    
private:
    void dfs(vector<vector<string>>& ret, vector<string>& elm, string& s, int pos)
    {
        int n = s.size();
        if (pos == n)
        {
            ret.push_back(elm);
            return;
        }
        for (int i = pos; i < n; ++i)
        {
            if(isP(s, pos, i))
            {
                elm.push_back(s.substr(pos, i-pos+1));
                dsf(ret, elm, s, i+1);
                elm.pop_back();
            }
        }
    }
    bool isP(string& str, int start, int end)
    {
        while (start < end && str[start] == str[end])
        {
            ++start;
            --end;
        }
        
        return (start >= end);
    }
};

File: leetcode/131.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> ret = new ArrayList<>();
        List<String> elm = new ArrayList<>();
        dfs(ret, elm, s, 0);
        return ret;
    }

    private void dfs(List<List<String>> ret, List<String> elm, String s, int pos) {
        int n = s.length();
        if (pos == n) {
            // List<String> e=new ArrayList<>(); //ok
            // e.addAll(elm);
            ret.add(new ArrayList<>(elm));
            return;
        }
        String cur;
        for (int i = pos + 1; i <= n; i++) { //i means end pos excusively
            if (isP(s, pos, i - 1)) {
                cur = s.substring(pos, i);
                elm.add(cur);
                dfs(ret, elm, s, i);
                elm.remove(elm.size() - 1);
            }
        }
    }

    private boolean isP(String s, int i, int j) {
        while (i < j) {
            if (s.charAt(i++) != s.charAt(j--)) {
                return false;
            }
        }
        return true;
    }
}